feat: add orogen task for genverter

eduardacoppo opened the pull request:
Depends on:
- [ ] https://github.com/tidewise/drivers-power_whisperpower/pull/7

# What is this PR for
<!-- A clear description of what this PR do and the changes made.
If you want you can also add the shortcut link here
- [ ] [Shortcut](http://) -->
[sc-70487]

This PR implements the PMGGenverter orogen task

# How I did it
<!-- Explain a little bit of your implementation -->

# Results, How I tested
<!-- Explain how you tested you PR, if there is a visual output,
     a GIF or image is welcome -->

# Checklist
- [x] Assign yourself  in the PR
- [x] Write unit tests (when relevant)
- [x] Run rubocop and rake tests locally

eduardacoppo committed:
feat: add orogen task for genverter

eduardacoppo committed:
feat: write task for genverter

eduardacoppo committed:
test: add tests for genverter task

higor-melo commented on test/pmg_genverter_task_test.rb:
```diff
@@ -0,0 +1,97 @@
+            )
+        end.to { have_one_new_sample task.full_status_port }
+
+        assert now <= status.time
+        assert status.time <= Time.now
```
This is new for me, also why do the asserts to verify that the sample is being received? Couldn't you just use different values?

eduardacoppo replied:
This is just asserting that the timestamp of the received sample falls within a reasonable range. That it's not from the future (`status.time <= Time.now`) and that the status time is not from before we wrote the sample (`now <= status.time`).

> why do the asserts to verify that the sample is being received?

The asserts are not verifying that the sample is being received. This is being done by the `have_one_new_sample` expectation.

> Couldn't you just use different values?

What values are you referring to? The message bits? How would this affect the test? The actual message contents are tested in the lib.

eduardacoppo committed:
feat: add port and control command implementation

eduardacoppo committed:
test: add tests for control_cmd

eduardacoppo committed:
feat: add  run time state port

eduardacoppo committed:
test: add tests for run time state port

seixasxbr committed:
feat: add device state port

seixasxbr committed:
feat: export device state type

seixasxbr committed:
feat: write generator state

doudou reviewed (commented):
Since we just discussed that the logic was wrong, I'm going to add this comment waiting for a new review request.

To keep things simple, I recommended that the component would send (false, false) for the required number of ms to "prime" the genset before actually passing commands every time there is a transition from false to true on the input command. Nothing more fancy than that.

seixasxbr committed:
fix: device startup logic

send a restart command for every transition of the control command
from 0 to 1, to 'prime' the device before actually passing commands
this commit also fix the tests

seixasxbr commented on tasks/PMGGenverterTask.cpp:
```diff
@@ -0,0 +1,118 @@
+            continue;
+        }
+
+        if (can_in.can_id == 0x204 || can_in.can_id == 0x205) {
+            m_ready_to_command = true;
+        }
```
from where this comes from?
btw this snippet will never be reached unless an message 0x205 is processed, because of the check before that one

jhonasiv replied:
Eduarda mentioned at the time that receiving messages indicates that the system can be commanded (it coudlnt before)

doudou replied:
It's actually not needed because it is what hasFullUpdate does.

In addition, we should process the control command only in the cycle where a full update is received, to avoid overloading the CAN bus. Something like

```
while (_can_in(...)) {
    <process>
    continue unless hasFullUpdate
    writeStates()
    handleControlCommand()
    resetFullUpdate()
}
```

I think it would make a lot more sense calling resetFullUpdate in this loop instead of inside the writeStates() call. Makes the overall logic clearer.

seixasxbr replied:
are you sure about that? the refresh rate of the 0x204 and 0x205 messages are 25 hz, and the rx message (control command) has the frequency of 50 hz.

seixasxbr replied:
just for the record, we decided to write if there is a 204 or 205 message

seixasxbr committed:
chore: make restart duration configurable

it also reorganize the code

jhonasiv commented on tasks/PMGGenverterTask.cpp:
```diff
@@ -0,0 +1,113 @@
+            m_restart_command_deadline = base::Time::now() + m_restart_duration;
+            m_last_command = control_cmd;
+        }
+        if (base::Time::now() <= m_restart_command_deadline) {
+            _can_out.write(m_driver.queryGeneratorCommand(false, false));
+        }
+        else {
+            _can_out.write(m_driver.queryGeneratorCommand(true, false));
+        }
```
I dont get this, you are restarting the generator every `m_restart_command_deadline` seconds?

jhonasiv replied:
Actually, I am not sure this is a restart, as you only write a start command (but not a stop), is it some sort of keepalive?

seixasxbr replied:
> I dont get this, you are restarting the generator every `m_restart_command_deadline` seconds?

I am restarting the device every time that there is a transition in the control command input port, from 0 to 1.
This is done sending the comand (false, false) for at least 200ms, according to the documentation.

seixasxbr replied:
> I dont get this, you are restarting the generator every `m_restart_command_deadline` seconds?

I am restarting the device every time that there is a transition in the control command input port, from 0 to 1.
This is done sending the comand (false, false) for at least 200ms, according to the documentation.

doudou commented on power_whisperpower.orogen:
```diff
@@ -45,6 +50,38 @@ task_context "DCPowerCubeTask" do
+    # The duration sending the restart command before the start command
+    #
+    # This is done to ensure that the device is fully reseted before the startup
+    property "restart_duration", "/base/Time"
```
For once, given that this value is determined by the device, I would have been OK with a hardcoded value.

But I'm fine with a property, just set it to the PMG value with some margin (e.g. 500ms) in the constructor.

doudou commented on tasks/PMGGenverterTask.cpp:
```diff
@@ -0,0 +1,113 @@
+        return GeneratorState::STOPPED;
+    }
+    else {
+        return GeneratorState::FAILURE;
```
?

Definitely not.

Failure is indicated by another field. Right now, I'm not so sure of the overall startup/stopping sequence but I'm guessing that combinations of the flags in `status` would indicate STARTING and STOPPING.

For the time being, I would check for RUNNING and return STOPPED in all other cases, and set failure if there are some alarm bit set.

doudou commented on tasks/PMGGenverterTask.cpp:
```diff
@@ -0,0 +1,113 @@
+}
+void PMGGenverterTask::handleControlCommand(bool control_cmd)
+{
+    if (control_cmd) {
```
Early return

```
if (!control_cmd) {
    _can_out.write(m_driver.queryGeneratorCommand(false, true));
    m_last_command = false;
    return;
}

if (!m_last_command) {
    <update deadline>
    m_last_command = true;
}

...
```

seixasxbr committed:
feat: set default restart duration property value at task constructor

seixasxbr committed:
fix: generator state parsing logic

seixasxbr committed:
chore: clear handle control command method

seixasxbr committed:
chore: suit to genset state struct refactor

seixasxbr committed:
fix: refactor updateHook

write at control command port only when there is a 0x204 or 0x205 message

jhonasiv commented on test/pmg_genverter_task_test.rb:
```diff
@@ -0,0 +1,217 @@
+        end
+    end
+
+    it "expects full_status to be reset and not outputted the message 0x205 is not "\
```
```suggestion
    it "expects full_status to be reset and not output when the message 0x205 is not "\
```

seixasxbr replied:
looks like outputted is correct either

doudou commented on power_whisperpower.orogen:
```diff
@@ -45,6 +50,38 @@ task_context "DCPowerCubeTask" do
+
+    # The duration sending the restart command before the start command
+    #
+    # This is done to ensure that the device is fully reseted before the startup
```
No need to fix that, but `reseted` does not exist, it's `reset` all the way (I have reset)

doudou commented on tasks/PMGGenverterTask.cpp:
```diff
@@ -0,0 +1,112 @@
+    }
+    if (!m_last_command) {
+        m_restart_command_deadline = base::Time::now() + m_restart_duration;
+        m_last_command = control_cmd;
```
Also not a hard ask, but since `control_cmd` here is known to be true ... I prefer making it explicit (that is, writing `m_last_command = true;`)
